{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Working with \"irregular\" Datafiles\n",
    "\n",
    "This notebook gives an example how to read \"irregular\" datafiles using `qpcr.Readers`.\n",
    "It makes use of the provided example data in the `Example Data` directory. \n",
    "\n",
    "\n",
    "#### Experimental background\n",
    "\n",
    "The corresponding experimental setup was as follows: \n",
    "Levels of Nonsense-mediated mRNA decay (NMD) sensitive (nmd) and insensitive (prot) transcript isoforms of HNRNPL and SRSF11 were measured by qPCR. As normalisers both 28S rRNA and Actin transcript levels were measured. The replicates are biological triplicates and technical douplicates. All measurements from the same qPCR sample were merged into hexaplicates (6 replicates). This was done in two separate HeLa cell lines (one with a specific gene knockout (KO), and one without (WT)), which were both treated to a plasmid-mediated rescue (+) or not (-), leading to four experimental conditions:\n",
    "\n",
    "\n",
    "| cell line \\\\ condition | rescue | no rescue |\n",
    "| ---------------------- | ------ | --------- |\n",
    "| knockout               | KO+    | KO-       |\n",
    "| wildtype               | WT+    | WT-       |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the qpcr module\n",
    "import qpcr\n",
    "from qpcr.Readers import SingleReader, MultiReader"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1 - Reading an \"irregular\" single-assay datafile\n",
    "---\n",
    "\n",
    "### 1.1 Setting up the `DataReader`\n",
    "First we set up the `qpcr.DataReader`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# our single-assay datafile\n",
    "file = \"./Example Data/Irregular Files/28S_irregular.csv\"\n",
    "\n",
    "# set up the reader\n",
    "reader = qpcr.DataReader()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 Specify how to read the file\n",
    "\n",
    "This datafile is an \"irregular\" file but not a \"Big Table\" file (hence it specifies different assays as separate tables). We know (you can check out the file for yourself, if you like) the assays have a \"header\" in the very first column of the file (so we know where to look for assays). \n",
    "\n",
    "#### The `assay_pattern`\n",
    "Also, we know they always follow the same kind of `pattern` (since the data was generated by QiaGen's Rotor-Gene, this \"assay header\" is always `\"Quantitative analysis of Cycling A.<color> (<my_assay>)\"`). The `qpcr.Parsers` have a way of using `regex` to find such assay headers and read the assay identifiers from them. We can either specify a `regex` pattern directly or use a pre-defined one using their `assay_pattern` method, which is available as an argument via the `DataReader`. For this case the pattern is already implemented and can be accessed through the key `\"Rotor-Gene\"`. \n",
    "\n",
    "#### Data columns\n",
    "We would also have to specify what the column names of the replicate identifiers and Ct values are. In this case the column names are `\"Name\"` for the identifiers and `\"Ct\"` for the Ct values (note that these are also the default that will be used if using the `DataReader`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning:\n",
      "It appears as if some provided kwargs were incompatible with pandas.read_csv()! Defaulting to standard settings for file-reading...\n",
      "If the kwargs you specified are actually important for file reading, try manually reading and parsing to avoid kwarg incompatibilities.\n"
     ]
    }
   ],
   "source": [
    "assay = reader.read(\n",
    "                        filename = file, \n",
    "\n",
    "                        # specify the assay pattern from the file\n",
    "                        assay_pattern = \"Rotor-Gene\",\n",
    "                        \n",
    "                        # specify the column where assay headers are \n",
    "                        # (this is the default)\n",
    "                        col = 0, \n",
    "\n",
    "                        # specify the column names of the data columns \n",
    "                        # (these are actually the defaults)\n",
    "                        id_label = \"Name\",\n",
    "                        ct_label = \"Ct\"\n",
    "                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'28S'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check what assay we got (it should be 28S)\n",
    "assay.id()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 Using the `qpcr.Readers.SingleReader` directly\n",
    "\n",
    "Of course, the `DataReader` did not actually read anything by itself. It simply looked at our provided specifics and decided to call on the `qpcr.Readers.SingleReader` to our file.  \n",
    "We could avoid the little computation required for Reader-inferencing by directly using the `qpcr.Readers.SingleReader`. \n",
    "\n",
    "Here's how:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first set up the reader and then use it's `pipe` method to read the data and get a `qpcr.Assay` object. In fact, this is just what the `DataReader` did for us before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning:\n",
      "It appears as if some provided kwargs were incompatible with pandas.read_csv()! Defaulting to standard settings for file-reading...\n",
      "If the kwargs you specified are actually important for file reading, try manually reading and parsing to avoid kwarg incompatibilities.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'28S'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reader = SingleReader()\n",
    "\n",
    "# read the file\n",
    "# this time we omit the specifics that are defaults anyway\n",
    "assay = reader.pipe(\n",
    "                        filename = file, \n",
    "                        # specify the assay pattern from the file\n",
    "                        assay_pattern = \"Rotor-Gene\",\n",
    "                )\n",
    "\n",
    "# and check if we still have 28S\n",
    "assay.id()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.4 Reading the Excel version\n",
    "So far we worked with `csv` files, but, of course, we could also have read an `excel` version of the file. To exemplify this, we read an irregular single-assay excel file containing an ActinB assay."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ACTB'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "excelfile = \"./Example Data/Irregular Files/actin_irregular.xlsx\"\n",
    "\n",
    "assay = reader.pipe(\n",
    "                        filename = excelfile, \n",
    "                        # specify the assay pattern from the file\n",
    "                        assay_pattern = \"Rotor-Gene\",\n",
    "                )\n",
    "\n",
    "# and check if we still have 28S\n",
    "assay.id()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2 - Reading \"irregular\" multi-assay files\n",
    "----\n",
    "What if we have multiple assays in the file? \n",
    "Well, we can easily do so using the `DataReader` for instance. To exemplify this we shall read a file that contains assays from 28S, and HNRNPL protein-coding and NMD-sensitive assays. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 Reading a single assay from a multi-assay datafile\n",
    "\n",
    "Extracting a single assay from a multi-assay datafile is pretty straight forward. All we need to do is to specify which assay we are interested in using the `assay` argument, which takes in the name of the assay."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning:\n",
      "It appears as if some provided kwargs were incompatible with pandas.read_csv()! Defaulting to standard settings for file-reading...\n",
      "If the kwargs you specified are actually important for file reading, try manually reading and parsing to avoid kwarg incompatibilities.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'28S'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_assay_file = \"./Example Data/Irregular Files/irregular_multi_assay.csv\"\n",
    "\n",
    "reader = qpcr.DataReader()\n",
    "\n",
    "assay = reader.read(\n",
    "                    filename = multi_assay_file, \n",
    "\n",
    "                    # specify the assay pattern from the file\n",
    "                    assay_pattern = \"Rotor-Gene\",\n",
    "\n",
    "                    # specify that we specifically want 28S for instance\n",
    "                    assay = \"28S\"\n",
    "                )\n",
    "\n",
    "# check what assay we got\n",
    "assay.id()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Reading multiple assays from a multi-assay datafile\n",
    "If we have a multi-assay datafile then we probably want to read all these assays, right? \n",
    "No problem for the `DataReader`. You may already have noticed that it's `read` method has an argument `multi_assay` which can be set to `True` if you wish to read multiple assays from your file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning:\n",
      "It appears as if some provided kwargs were incompatible with pandas.read_csv()! Defaulting to standard settings for file-reading...\n",
      "If the kwargs you specified are actually important for file reading, try manually reading and parsing to avoid kwarg incompatibilities.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "([<qpcr.Assay at 0x12af1a9d0>,\n",
       "  <qpcr.Assay at 0x12af1ac40>,\n",
       "  <qpcr.Assay at 0x12b0c3310>],\n",
       " [])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assays = reader.read(\n",
    "                    filename = multi_assay_file, \n",
    "\n",
    "                    # specify that we want multiple assays\n",
    "                    multi_assay = True,\n",
    "\n",
    "                    # specify the assay pattern from the file\n",
    "                    assay_pattern = \"Rotor-Gene\",\n",
    "\n",
    "                    # we need to reset the Reader that DataReader uses \n",
    "                    # as it currently thinks we are interested in \n",
    "                    # single assays (this is mainly a required technicality)\n",
    "                    reset = True\n",
    "                )\n",
    "\n",
    "# check what assays we got\n",
    "assays"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, you may notice that `two lists` are actually returned now, one containing three assays, the other one being empty. Why is that? \n",
    "\n",
    "Well, the two lists are for `\"assays-of-interest\"` and `\"normaliser-assays\"`. However, since the Reader has no way of knowing which of our assays is actually a normaliser an which is not, all assays are assigned to the \"assays-of-interest\" list. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 - Reading \"assays\" and \"normalisers\" from multi-assay files\n",
    "Well wouldn't it be nice if we could let the Reader somehow know which of our assays are actually normalisers? Luckily, we can do just that super easily by adding `decorators` to the assays in our datafile. \n",
    "\n",
    "Check out [the documentation about decorators](https://noahhenrikkleinschmidt.github.io/qpcr/Parsers.html#decorators) to learn how to do this. \n",
    "\n",
    "> Note\n",
    ">\n",
    "> Once we have decorated our assays we can even omit the `assay_pattern` argument if we want to, since the Reader (or, more acurately, the Parser that is being called by the Reader to extract the data) now knows what to look for (i.e. the decorators)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([<qpcr.Assay at 0x12af1adf0>, <qpcr.Assay at 0x12b109fd0>],\n",
       " [<qpcr.Assay at 0x12af408e0>])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decorated_multi_assay_file = \"./Example Data/Irregular Files/irregular_multi_assay_decorated.xlsx\"\n",
    "\n",
    "\n",
    "assays = reader.read(\n",
    "                    filename = decorated_multi_assay_file, \n",
    "\n",
    "                    # specify that we want multiple assays\n",
    "                    multi_assay = True,\n",
    "\n",
    "                    # specify that the assays are decorated\n",
    "                    decorator = True,\n",
    "\n",
    "                    # specify the assay pattern from the file\n",
    "                    assay_pattern = \"Rotor-Gene\",\n",
    "                )\n",
    "\n",
    "# check what assays we got\n",
    "assays"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice how the first list now only has two `qpcr.Assay` objects while the second one now has one? This is because the Reader now knew that 28S was actually supposed to be a normaliser and not an assay-of-interest.\n",
    "\n",
    "> Note\n",
    ">\n",
    "> Boolean `decorator` arguments only work for multi_assay files. If wish to use decorators to get specific assays you have to specify an actual decorator key! The save alternative to `decorator = True` is `decorator = \"qpcr:all\"` which will have the same effect."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.5 - Extracting only specific assays from a multi-assay file\n",
    "If you only wish extract specific assays you can either just read them all and then discard those you don't care about (this probably most straightforward), or decorate them and specify which decorator to extract.\n",
    "\n",
    "In fact, using a `Reader` you will *always* be extracting both assays and normalisers if possible. \n",
    "\n",
    "In case you really only want to read some specific assays you can use the `qpcr.Parsers` directly."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we will only extract the as \"assays\" decorated datasets from the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning:\n",
      "It appears as if some provided kwargs were incompatible with pandas.read_excel()! Defaulting to standard settings for file-reading...\n",
      "If the kwargs you specified are actually important for file reading, try manually reading and parsing to avoid kwarg incompatibilities.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dict_keys(['HNRNPL protein', 'HNRNPL NMD'])"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from qpcr.Parsers import ExcelParser\n",
    "\n",
    "# setup the Excel Parser\n",
    "parser = ExcelParser()\n",
    "\n",
    "# define the assay pattern\n",
    "parser.assay_pattern(\"Rotor-Gene\")\n",
    "\n",
    "\n",
    "# specify that we are only interested \n",
    "# in as assay decorated datasets\n",
    "decorator = \"qpcr:assay\"\n",
    "\n",
    "# pipe the file through the parser\n",
    "assays = parser.pipe( \n",
    "                        filename = decorated_multi_assay_file,\n",
    "\n",
    "                        # specify what decorators we are interested in\n",
    "                        decorator = decorator\n",
    "                )\n",
    "\n",
    "\n",
    "# now we have a dictionary of assay \n",
    "# identifiers and their dataframes\n",
    "assays.keys()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now convert the dictionary to `qpcr.Assay` objects by:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<qpcr.Assay at 0x12b267190>, <qpcr.Assay at 0x12b12da00>]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assays = [ qpcr.Assay(df, id) for id, df in assays.items() ]\n",
    "\n",
    "assays"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "652a90e61eb9c1f97ec337805654fe3aa8651960163f76bc5e0407c363634b53"
  },
  "kernelspec": {
   "display_name": "Python 3.8.12 64-bit ('base': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
