{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Working with \"Big Table\" Datafiles\n",
    "\n",
    "This notebook gives an example how to read \"Big Table\" datafiles using `qpcr.Readers`.\n",
    "It makes use of the provided example data in the `Example Data` directory. \n",
    "\n",
    "\n",
    "#### Experimental background\n",
    "\n",
    "The corresponding experimental setup was as follows: \n",
    "Levels of Nonsense-mediated mRNA decay (NMD) sensitive (nmd) and insensitive (prot) transcript isoforms of HNRNPL and SRSF11 were measured by qPCR. As normalisers both 28S rRNA and Actin transcript levels were measured. The replicates are biological triplicates and technical douplicates. All measurements from the same qPCR sample were merged into hexaplicates (6 replicates). This was done in two separate HeLa cell lines (one with a specific gene knockout (KO), and one without (WT)), which were both treated to a plasmid-mediated rescue (+) or not (-), leading to four experimental conditions:\n",
    "\n",
    "\n",
    "| cell line \\\\ condition | rescue | no rescue |\n",
    "| ---------------------- | ------ | --------- |\n",
    "| knockout               | KO+    | KO-       |\n",
    "| wildtype               | WT+    | WT-       |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the qpcr module\n",
    "import qpcr\n",
    "from qpcr.Readers import BigTableReader"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1 - Reading a \"vertical\" Big Table File\n",
    "---\n",
    "\n",
    "### 1.1 Setting up the `DataReader`\n",
    "First we set up the `qpcr.DataReader`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# our single-assay datafile\n",
    "file = \"./Example Data/Big Table Files/vertical_bigtable_decorated.csv\"\n",
    "\n",
    "# set up the reader\n",
    "reader = qpcr.DataReader()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 Specify how to read the file\n",
    "\n",
    "This datafile is a \"vertical\" Big Table file. That means it stores assays above each other and has an `assay`, a `sample` (groups), and `Ct` column. In fact, this example datafile has nothing else in it, but reading a file that specifies other stuff would work just the same. \n",
    "\n",
    "In order to read Big Table file we need to specify `big_table = True` for the `DataReader` or immediately use the `qpcr.Readers.BigTableReader`.\n",
    "\n",
    "#### The `kind` parameter\n",
    "One important parameter we have to specify is the architecture of the table. The table in our file is \"vertical\" so we need to specify `kind = \"vertical\"` in order to tell the `Reader` how to interpret the data it reads. The `kind` parameter tells the `BigTableReader` which parsing method to use to extract the individual assays. The other option here would correspondingly be `kind = \"horizontal\"`.\n",
    "\n",
    "Also, note how the filename already has \"decorated\" in it? That's because the file already includes a `@qpcr` decorator column specifying which assays are \"assays-of-interest\" and which ones are normalisers. \n",
    "\n",
    "Like this we can read the file and immediately pass it on to our analysis.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "assay, normalisers = reader.read(\n",
    "                                        filename = file, \n",
    "\n",
    "                                        # specify the big table\n",
    "                                        big_table = True,\n",
    "\n",
    "                                        # specify that the data is decorated\n",
    "                                        decorator = True,\n",
    "\n",
    "                                        # specify the kind of big table\n",
    "                                        kind = \"vertical\",\n",
    "\n",
    "                                        # specify which columns store\n",
    "                                        # the relevant data\n",
    "                                        assay_col = \"Assay\", \n",
    "                                        id_col = \"Name\",\n",
    "                                        ct_col = \"Ct\"\n",
    "                        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 Using the `qpcr.Readers.BigTableReader` directly\n",
    "\n",
    "Of course, we can, again, also use the `BigTableReader` directly. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "reader = BigTableReader()\n",
    "\n",
    "# read the file\n",
    "assays, normalisers = reader.pipe(\n",
    "                                        filename = file, \n",
    "\n",
    "                                        # specify that the data is decorated\n",
    "                                        decorator = True,\n",
    "\n",
    "                                        # specify the kind of big table\n",
    "                                        kind = \"vertical\",\n",
    "\n",
    "                                        # specify which columns store\n",
    "                                        # the relevant data\n",
    "                                        assay_col = \"Assay\", \n",
    "                                        id_col = \"Name\",\n",
    "                                        ct_col = \"Ct\"\n",
    "                                )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2 Reading a \"horizontal\" Big Table File\n",
    "----------------------------------------------------------------\n",
    "\n",
    "Reading a \"horizontal\" big table file is actually much more tricky than it's vertical counterpart. Why's that? Well, because in \"horizontal\" big table files the replicates are aligned in columns next to each other, and they usually have different column headers as well. That means we can no longer rely on a single column or even a specific `regex` pattern to extract our values. Instead we need to rely on `qpcr decorators` to help guide the parsing algorithm. To this end, we have the `@qpcr:group` decorator, which has to be placed _in the cell immediately above the first replicate column of each group_. Sounds too abstract? Check out the the `Getting Started` notebook again and look at the example there (it's actually not complicated at all).  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 - Using the `BigTableReader`\n",
    "\n",
    "We will actually now read a datafile that is not one of our normal example files. It's a file from [Yang et al. (2018)](https://doi.org/10.5061/dryad.3274k) who analysed the expression of micro-RNAs (miRNAs) in alvular interstitial cells of dogs suffering from Canine myxomatous mitral valve disease. The architecture of the datafile they uploaded to Zenodo is a \"horizontal\" big table containing assays of miRNAs on separate rows and replicate Ct values from \"normal\" (healthy), \"mildly\" ill, or \"severely\" ill dogs as three replicate groups of five replicates each. \n",
    "\n",
    "Regrettably, they did not include the Ct values from the normaliser assays they used, so we will be unable to to anything useful beyond \"file reading\".\n",
    "\n",
    "> Please, note that the datafile we are reading is actually a processed version of the original file they uploaded that merged the separate sheets from the excel file into a single-sheet big table. Also, duplicate assays were dropped for the purpose of reading. Currently, the `BigTableReader` in `\"horizontal\"` parsing mode only supports *uniquely* labelled assays (so if there _are_ duplicate assays they must be differently labelled!)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Warning:\n",
      "It appears as if some provided kwargs were incompatible with pandas.read_csv()! Defaulting to standard settings for file-reading...\n",
      "If the kwargs you specified are actually important for file reading, try manually reading and parsing to avoid kwarg incompatibilities.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[<qpcr.Assay at 0x127f62850>,\n",
       " <qpcr.Assay at 0x11352eeb0>,\n",
       " <qpcr.Assay at 0x1135019d0>,\n",
       " <qpcr.Assay at 0x11351d340>,\n",
       " <qpcr.Assay at 0x127f47eb0>,\n",
       " <qpcr.Assay at 0x127f6e940>,\n",
       " <qpcr.Assay at 0x127f6e640>,\n",
       " <qpcr.Assay at 0x113525400>,\n",
       " <qpcr.Assay at 0x113549400>,\n",
       " <qpcr.Assay at 0x11353ad90>,\n",
       " <qpcr.Assay at 0x1135327f0>,\n",
       " <qpcr.Assay at 0x127f17c40>,\n",
       " <qpcr.Assay at 0x127fe3670>,\n",
       " <qpcr.Assay at 0x113520ee0>,\n",
       " <qpcr.Assay at 0x127fe61f0>,\n",
       " <qpcr.Assay at 0x127fe6a30>,\n",
       " <qpcr.Assay at 0x127fe6970>,\n",
       " <qpcr.Assay at 0x127fe6e80>,\n",
       " <qpcr.Assay at 0x127fe67c0>,\n",
       " <qpcr.Assay at 0x127fe3550>,\n",
       " <qpcr.Assay at 0x127fe3100>,\n",
       " <qpcr.Assay at 0x113525550>,\n",
       " <qpcr.Assay at 0x11352eac0>,\n",
       " <qpcr.Assay at 0x127fea4f0>,\n",
       " <qpcr.Assay at 0x127fea7f0>,\n",
       " <qpcr.Assay at 0x127fead00>,\n",
       " <qpcr.Assay at 0x127fdba30>,\n",
       " <qpcr.Assay at 0x128012a00>,\n",
       " <qpcr.Assay at 0x128012f70>,\n",
       " <qpcr.Assay at 0x1280120a0>,\n",
       " <qpcr.Assay at 0x1280129a0>,\n",
       " <qpcr.Assay at 0x128017fd0>,\n",
       " <qpcr.Assay at 0x113549670>,\n",
       " <qpcr.Assay at 0x128017280>,\n",
       " <qpcr.Assay at 0x1280170a0>,\n",
       " <qpcr.Assay at 0x127fea9a0>,\n",
       " <qpcr.Assay at 0x127f85160>,\n",
       " <qpcr.Assay at 0x128012190>,\n",
       " <qpcr.Assay at 0x127fdbc70>,\n",
       " <qpcr.Assay at 0x127fe6fa0>,\n",
       " <qpcr.Assay at 0x127fe3970>,\n",
       " <qpcr.Assay at 0x127fe3f10>,\n",
       " <qpcr.Assay at 0x1280204f0>,\n",
       " <qpcr.Assay at 0x128025d30>,\n",
       " <qpcr.Assay at 0x1280257f0>,\n",
       " <qpcr.Assay at 0x128025280>,\n",
       " <qpcr.Assay at 0x128025940>,\n",
       " <qpcr.Assay at 0x113525040>,\n",
       " <qpcr.Assay at 0x1134e3e20>,\n",
       " <qpcr.Assay at 0x113549790>,\n",
       " <qpcr.Assay at 0x127fea6a0>,\n",
       " <qpcr.Assay at 0x127fdbe80>,\n",
       " <qpcr.Assay at 0x1280125b0>,\n",
       " <qpcr.Assay at 0x128029f70>,\n",
       " <qpcr.Assay at 0x1280291f0>,\n",
       " <qpcr.Assay at 0x128029b50>,\n",
       " <qpcr.Assay at 0x128031df0>,\n",
       " <qpcr.Assay at 0x1280317f0>,\n",
       " <qpcr.Assay at 0x1280313a0>,\n",
       " <qpcr.Assay at 0x11356bf70>,\n",
       " <qpcr.Assay at 0x128031610>,\n",
       " <qpcr.Assay at 0x128029be0>,\n",
       " <qpcr.Assay at 0x128029880>,\n",
       " <qpcr.Assay at 0x128031b20>,\n",
       " <qpcr.Assay at 0x128012fa0>,\n",
       " <qpcr.Assay at 0x128031eb0>,\n",
       " <qpcr.Assay at 0x128033f40>,\n",
       " <qpcr.Assay at 0x1280334f0>,\n",
       " <qpcr.Assay at 0x128033ca0>,\n",
       " <qpcr.Assay at 0x12803e760>,\n",
       " <qpcr.Assay at 0x12803e9d0>,\n",
       " <qpcr.Assay at 0x1135254c0>,\n",
       " <qpcr.Assay at 0x113549b80>,\n",
       " <qpcr.Assay at 0x11353a220>,\n",
       " <qpcr.Assay at 0x128044730>,\n",
       " <qpcr.Assay at 0x12803ed90>,\n",
       " <qpcr.Assay at 0x12803edf0>,\n",
       " <qpcr.Assay at 0x12803e550>,\n",
       " <qpcr.Assay at 0x1280d4bb0>,\n",
       " <qpcr.Assay at 0x12803ef10>,\n",
       " <qpcr.Assay at 0x1280315b0>,\n",
       " <qpcr.Assay at 0x128012b20>,\n",
       " <qpcr.Assay at 0x128046070>,\n",
       " <qpcr.Assay at 0x128046e80>,\n",
       " <qpcr.Assay at 0x128046700>,\n",
       " <qpcr.Assay at 0x11336fee0>,\n",
       " <qpcr.Assay at 0x11336fa90>,\n",
       " <qpcr.Assay at 0x11336f070>,\n",
       " <qpcr.Assay at 0x11336f370>,\n",
       " <qpcr.Assay at 0x1280e4a60>,\n",
       " <qpcr.Assay at 0x128046970>,\n",
       " <qpcr.Assay at 0x128033c10>,\n",
       " <qpcr.Assay at 0x1280258b0>,\n",
       " <qpcr.Assay at 0x1280d4cd0>,\n",
       " <qpcr.Assay at 0x113377b80>,\n",
       " <qpcr.Assay at 0x113377970>,\n",
       " <qpcr.Assay at 0x113377190>,\n",
       " <qpcr.Assay at 0x113381e20>,\n",
       " <qpcr.Assay at 0x113381220>,\n",
       " <qpcr.Assay at 0x113381190>,\n",
       " <qpcr.Assay at 0x113385f40>,\n",
       " <qpcr.Assay at 0x113385af0>,\n",
       " <qpcr.Assay at 0x113385e50>,\n",
       " <qpcr.Assay at 0x1280f3d60>,\n",
       " <qpcr.Assay at 0x113385d30>,\n",
       " <qpcr.Assay at 0x1280335b0>,\n",
       " <qpcr.Assay at 0x1280338e0>,\n",
       " <qpcr.Assay at 0x12803e3d0>,\n",
       " <qpcr.Assay at 0x113385dc0>,\n",
       " <qpcr.Assay at 0x11336ff40>,\n",
       " <qpcr.Assay at 0x113388700>,\n",
       " <qpcr.Assay at 0x1133889d0>,\n",
       " <qpcr.Assay at 0x113388520>,\n",
       " <qpcr.Assay at 0x1133947f0>,\n",
       " <qpcr.Assay at 0x113394c40>,\n",
       " <qpcr.Assay at 0x1133948b0>,\n",
       " <qpcr.Assay at 0x113388820>,\n",
       " <qpcr.Assay at 0x113394b20>,\n",
       " <qpcr.Assay at 0x113388b80>,\n",
       " <qpcr.Assay at 0x113394700>,\n",
       " <qpcr.Assay at 0x113381400>,\n",
       " <qpcr.Assay at 0x1133983a0>,\n",
       " <qpcr.Assay at 0x113398be0>,\n",
       " <qpcr.Assay at 0x113398910>,\n",
       " <qpcr.Assay at 0x1133989a0>,\n",
       " <qpcr.Assay at 0x1133a0e20>,\n",
       " <qpcr.Assay at 0x1133a0760>,\n",
       " <qpcr.Assay at 0x1133a0970>,\n",
       " <qpcr.Assay at 0x1133a1f10>,\n",
       " <qpcr.Assay at 0x1133a1ca0>,\n",
       " <qpcr.Assay at 0x1280fd7f0>,\n",
       " <qpcr.Assay at 0x1133a0a00>,\n",
       " <qpcr.Assay at 0x1133a1fa0>,\n",
       " <qpcr.Assay at 0x1133a17f0>,\n",
       " <qpcr.Assay at 0x1133a0bb0>,\n",
       " <qpcr.Assay at 0x113394400>,\n",
       " <qpcr.Assay at 0x113394760>,\n",
       " <qpcr.Assay at 0x1133853a0>,\n",
       " <qpcr.Assay at 0x1133a7760>,\n",
       " <qpcr.Assay at 0x1133a7c70>,\n",
       " <qpcr.Assay at 0x1133a7c10>,\n",
       " <qpcr.Assay at 0x1134b5d90>,\n",
       " <qpcr.Assay at 0x1134b57f0>,\n",
       " <qpcr.Assay at 0x1134b5c70>,\n",
       " <qpcr.Assay at 0x1134b5580>,\n",
       " <qpcr.Assay at 0x1133a7e20>,\n",
       " <qpcr.Assay at 0x1133a7790>,\n",
       " <qpcr.Assay at 0x1133857c0>,\n",
       " <qpcr.Assay at 0x113394670>,\n",
       " <qpcr.Assay at 0x1134b7190>,\n",
       " <qpcr.Assay at 0x1133a7d90>,\n",
       " <qpcr.Assay at 0x1134b7f10>,\n",
       " <qpcr.Assay at 0x1134b7eb0>,\n",
       " <qpcr.Assay at 0x1134b7c70>,\n",
       " <qpcr.Assay at 0x1134bed90>,\n",
       " <qpcr.Assay at 0x1134beb50>,\n",
       " <qpcr.Assay at 0x1134be490>,\n",
       " <qpcr.Assay at 0x128113070>,\n",
       " <qpcr.Assay at 0x1134b7d60>,\n",
       " <qpcr.Assay at 0x1134b7490>,\n",
       " <qpcr.Assay at 0x1134b7730>,\n",
       " <qpcr.Assay at 0x1134b7a00>,\n",
       " <qpcr.Assay at 0x1134be970>,\n",
       " <qpcr.Assay at 0x1133a7430>,\n",
       " <qpcr.Assay at 0x1134c5be0>,\n",
       " <qpcr.Assay at 0x1134c5fd0>,\n",
       " <qpcr.Assay at 0x1134c5b80>,\n",
       " <qpcr.Assay at 0x1134cddf0>,\n",
       " <qpcr.Assay at 0x1134cdfa0>,\n",
       " <qpcr.Assay at 0x128103ac0>,\n",
       " <qpcr.Assay at 0x1134d2100>,\n",
       " <qpcr.Assay at 0x1134d2d90>,\n",
       " <qpcr.Assay at 0x1134d2be0>,\n",
       " <qpcr.Assay at 0x1134d2820>,\n",
       " <qpcr.Assay at 0x113398af0>,\n",
       " <qpcr.Assay at 0x1133a00a0>,\n",
       " <qpcr.Assay at 0x1134be400>,\n",
       " <qpcr.Assay at 0x1134beac0>,\n",
       " <qpcr.Assay at 0x113388460>,\n",
       " <qpcr.Assay at 0x1134d5d90>,\n",
       " <qpcr.Assay at 0x1134d5280>,\n",
       " <qpcr.Assay at 0x1134d57c0>,\n",
       " <qpcr.Assay at 0x1134dfdf0>,\n",
       " <qpcr.Assay at 0x1134df280>,\n",
       " <qpcr.Assay at 0x1134df8e0>,\n",
       " <qpcr.Assay at 0x1134d5d60>,\n",
       " <qpcr.Assay at 0x1134d5a60>,\n",
       " <qpcr.Assay at 0x1134dfa60>,\n",
       " <qpcr.Assay at 0x1134d5b80>,\n",
       " <qpcr.Assay at 0x1134c56d0>,\n",
       " <qpcr.Assay at 0x1134c5a00>,\n",
       " <qpcr.Assay at 0x1134d5e20>,\n",
       " <qpcr.Assay at 0x1134e35e0>,\n",
       " <qpcr.Assay at 0x1134e3370>,\n",
       " <qpcr.Assay at 0x1134ec910>,\n",
       " <qpcr.Assay at 0x1134ecd00>,\n",
       " <qpcr.Assay at 0x1134ecb50>,\n",
       " <qpcr.Assay at 0x1134ec970>,\n",
       " <qpcr.Assay at 0x1134eefd0>,\n",
       " <qpcr.Assay at 0x1134ee700>,\n",
       " <qpcr.Assay at 0x1134ee4f0>,\n",
       " <qpcr.Assay at 0x1134eecd0>,\n",
       " <qpcr.Assay at 0x1134d2ac0>,\n",
       " <qpcr.Assay at 0x1134eee50>,\n",
       " <qpcr.Assay at 0x1134ecd60>,\n",
       " <qpcr.Assay at 0x1134e3640>,\n",
       " <qpcr.Assay at 0x113388730>,\n",
       " <qpcr.Assay at 0x1134e3f10>,\n",
       " <qpcr.Assay at 0x1134f3790>,\n",
       " <qpcr.Assay at 0x1134df670>,\n",
       " <qpcr.Assay at 0x1134fdd90>,\n",
       " <qpcr.Assay at 0x1134fd880>,\n",
       " <qpcr.Assay at 0x1134fd520>,\n",
       " <qpcr.Assay at 0x1134fdd60>,\n",
       " <qpcr.Assay at 0x1134fd250>,\n",
       " <qpcr.Assay at 0x1134f3c10>,\n",
       " <qpcr.Assay at 0x1134f3cd0>,\n",
       " <qpcr.Assay at 0x1134eed00>,\n",
       " <qpcr.Assay at 0x1134c53d0>,\n",
       " <qpcr.Assay at 0x128149f70>,\n",
       " <qpcr.Assay at 0x1134fda30>,\n",
       " <qpcr.Assay at 0x113501ac0>,\n",
       " <qpcr.Assay at 0x113501c10>,\n",
       " <qpcr.Assay at 0x11350ad90>,\n",
       " <qpcr.Assay at 0x11350a5e0>,\n",
       " <qpcr.Assay at 0x11350ab20>,\n",
       " <qpcr.Assay at 0x113510d30>,\n",
       " <qpcr.Assay at 0x113510d60>,\n",
       " <qpcr.Assay at 0x12815b850>,\n",
       " <qpcr.Assay at 0x11350aa00>,\n",
       " <qpcr.Assay at 0x113510df0>,\n",
       " <qpcr.Assay at 0x1134d5610>,\n",
       " <qpcr.Assay at 0x113510f40>,\n",
       " <qpcr.Assay at 0x1134fdf40>,\n",
       " <qpcr.Assay at 0x1134dfbe0>,\n",
       " <qpcr.Assay at 0x1135135e0>,\n",
       " <qpcr.Assay at 0x1135138e0>,\n",
       " <qpcr.Assay at 0x113513160>,\n",
       " <qpcr.Assay at 0x113513f40>,\n",
       " <qpcr.Assay at 0x11351de20>,\n",
       " <qpcr.Assay at 0x128163f40>,\n",
       " <qpcr.Assay at 0x11351d8e0>,\n",
       " <qpcr.Assay at 0x11351d6a0>,\n",
       " <qpcr.Assay at 0x113513a00>,\n",
       " <qpcr.Assay at 0x113513df0>,\n",
       " <qpcr.Assay at 0x11350a9a0>,\n",
       " <qpcr.Assay at 0x1135107c0>,\n",
       " <qpcr.Assay at 0x11350a040>,\n",
       " <qpcr.Assay at 0x113520cd0>,\n",
       " <qpcr.Assay at 0x1135200d0>,\n",
       " <qpcr.Assay at 0x1135108b0>,\n",
       " <qpcr.Assay at 0x113525df0>,\n",
       " <qpcr.Assay at 0x113525460>,\n",
       " <qpcr.Assay at 0x1135207f0>,\n",
       " <qpcr.Assay at 0x11352ebb0>,\n",
       " <qpcr.Assay at 0x11352ef70>,\n",
       " <qpcr.Assay at 0x11352e2e0>,\n",
       " <qpcr.Assay at 0x113525d00>,\n",
       " <qpcr.Assay at 0x11352ec40>,\n",
       " <qpcr.Assay at 0x113525dc0>,\n",
       " <qpcr.Assay at 0x113520be0>,\n",
       " <qpcr.Assay at 0x1134ee0d0>,\n",
       " <qpcr.Assay at 0x11352e430>,\n",
       " <qpcr.Assay at 0x113520e50>,\n",
       " <qpcr.Assay at 0x113532ee0>,\n",
       " <qpcr.Assay at 0x113532e20>,\n",
       " <qpcr.Assay at 0x113532c70>,\n",
       " <qpcr.Assay at 0x11353a2e0>,\n",
       " <qpcr.Assay at 0x11353a670>,\n",
       " <qpcr.Assay at 0x11353a6a0>,\n",
       " <qpcr.Assay at 0x11353aee0>,\n",
       " <qpcr.Assay at 0x11353a580>,\n",
       " <qpcr.Assay at 0x113501d00>,\n",
       " <qpcr.Assay at 0x113520b20>,\n",
       " <qpcr.Assay at 0x11352e490>,\n",
       " <qpcr.Assay at 0x127f6ee80>,\n",
       " <qpcr.Assay at 0x127f62be0>,\n",
       " <qpcr.Assay at 0x127f17280>,\n",
       " <qpcr.Assay at 0x11352ee80>,\n",
       " <qpcr.Assay at 0x113540580>,\n",
       " <qpcr.Assay at 0x1135408e0>,\n",
       " <qpcr.Assay at 0x113540640>,\n",
       " <qpcr.Assay at 0x128194070>,\n",
       " <qpcr.Assay at 0x113549970>,\n",
       " <qpcr.Assay at 0x113549c70>,\n",
       " <qpcr.Assay at 0x127f85040>]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = \"./Example Data/Big Table Files/horizontal_bigtable_decorated.csv\"\n",
    "\n",
    "# we need to reset the reader, because \n",
    "# it currently still stores the assays \n",
    "# from the vertical big table file...\n",
    "reader.clear()\n",
    "\n",
    "assays, normalisers = reader.pipe(\n",
    "                                    filename = file,\n",
    "\n",
    "                                    # we specify that it's a horizontal table\n",
    "                                    kind = \"horizontal\",\n",
    "\n",
    "                                    # we **have to** specify the \n",
    "                                    # number of replicates per group\n",
    "                                    replicates = 5,\n",
    "\n",
    "                                    # we also specify the group names (because they \n",
    "                                    # will not be inferrable due to the different \n",
    "                                    # column names of the replicates)\n",
    "                                    names = [\"normal\", \"mild\", \"severe\"],\n",
    "                                    \n",
    "                                    # we must also specify which column specifies the assays\n",
    "                                    # NOTE: in this mode this is handled by the `id_col` argument!\n",
    "                                    id_col = \"Target\"\n",
    "                                \n",
    "                                )\n",
    "\n",
    "print( f\"Loaded assays: {len(assays)}\" ) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And just like this we have read 286 new `qpcr.Assay`s that we can now either process through native `qpcr` methods, or we can use the data we just assembled with external tools such as `pandas`, `scipy`, or `numpy`. After all, there is no need to restrict ourselves to using `qpcr` only for its developed \"core business\" of Delta-Delta-Ct analysis. We can use the data `qpcr` reads for any kind of analysis we want."
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "652a90e61eb9c1f97ec337805654fe3aa8651960163f76bc5e0407c363634b53"
  },
  "kernelspec": {
   "display_name": "Python 3.8.12 64-bit ('base': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
